### PR TITLE
Try/even better previews

### PR DESCRIPTION
**Pull request by retrofox**
This PR has been cloned from https://github.com/WordPress/gutenberg/pull/16113 in order to achieve the same goal but from a different approach.

In short, the final idea behind of these PRs is improving the visualization of the `<BlockPreview />`.

### Some screenshots

**<Button />**

**Preview styles in the sidebar**

**Full preview**

## How can use it

Unlike the previous approach, this proposal does not require that the dimensions of the block (`srcWidth` and `srcHeight`), the scale factor be defined (`scaleFactor`), and whether or not the scale should be adjusted (`fitToScale`). So, we only need to pass the `blocks` property to the `<BlockPreview />` component to make it work.

```jsx
<BlockPreviewContent blocks={ cloneBlock( block, { className: styleClassName } ) } />
</div>
```

### Width and the Scale.

The `scale` is computed **only** considering the width of the blocks and its container. We could polish this approach incorporating the possibility to fit the preview in height, centered, etc. I image something like the `background-size` and `background-position`.

The scaled is computed dividing the width of the container element (it belongs to the `<PreviewBlock />` component) by the width of the block to visualize. **VERY IMPORTANT: it works with only one block so far**. However, I'm quite confident that we can make it work for more many blocks as well.

```
const scale = containerWidth / blocksWidth;
```

To get the width of the container we use the `useRef()` hook and the `ref` property. Kind simple.

Getting the width of the block to preview is kinda hacky at the moment. There is something funny when we try to get the DOM element. Although we are using the `useLayoutEffect()` hook, the dom isn't there yet meaning that when we try to pick it up through its clientId (using a dom/util function) it returns null.
First hack: setTimeout(). Wrapping the code into this function we can get the DOM element. I'd like to think that we can get a better solution, digging the reason why it happens. It seems to be related to the `<BlockEditorProvider />` block. I've followed advice from @youknowriad, creating a new component to wrap the `<BlockList />` but didn't work either. Maybe I'm missing something.

Another relevant change that I'm introducing here is using a CSS class to identify what part(elements) of the Block we want to visualize. AFAIK, This is completely needed since we don't have any other way to detect the exact part of the component that we want.

Although the `getBlockDOMNode()` returns the DOM Element of the block container, generally it doesn't match which what we'd like to visualize. And we can't generalize among all blocks saying ... _just pick the first child of its children_, or something like that.
Adding this new `.wp-block-preview-container` to those elements of the blocks, we can run a query selection easily. For this, I've added the `getClockPreviewContainerDOMNode()` function.
It means that we need to explicitly add this class the blocks that can be previewed. This PR adds the class to the `table`, `button` and `pullquote` component. While this seems to be complicated, it could also be considered one that allows us to easily define which elements we want to visualize from the block.

Finally, once we have the container width and the block with, the rest is a matter of simple calculation.

# How to test

* create a post with a `table`, `pullquote` and/or a `button`.
* check how looks the preview in the different locations: Sidebar, Switcher and the Full Preview Popover.

# Status
Pretty sure there are a lot of things to improve/change and at the end, I'm not 100% confident about this approach neither.

g-preview

### CODE REVIEW

File: packages/block-editor/src/components/block-preview/index.js [hunk 1]
```diff
@@ -33,19 +37,57 @@ function BlockPreview( props ) {
...
+	const previewRef = useRef( null );
+
+	const [ previewScale, setPreviewScale ] = useState( 1 );
+	const [ visibility, setVisibility ] = useState( 'hidden' );
+
+	let timerId;
+
+	// Dynamically calculate the scale factor
+	useLayoutEffect( () => {
+		const { clientId } = blocks;
```
Comment (getdave) [inline packages/block-editor/src/components/block-preview/index.js]: `blocks` can be an array. This approach won't work in that situation.

File: packages/block-editor/src/components/block-preview/index.js [hunk 2]
```diff
@@ -33,19 +37,57 @@ function BlockPreview( props ) {
...
+			const previewBlocksWidth = previewDomElements ? previewDomElements.offsetWidth : containerWidth;
+			const scale = Math.min( containerWidth / previewBlocksWidth ) || 1;
+
+			setPreviewScale( scale );
+			setVisibility( 'visible' );
+		}, 10 );
+	} );
+
+	useEffect( () => {
+		return () => window.clearTimeout( timerId );
```
Comment (getdave) [inline packages/block-editor/src/components/block-preview/index.js]: Why are you clearing the timer here and not in the layout effect? Genuinely curious.

I did this by returning a function from the `useLayoutEffect` which cleared the timer.

File: packages/block-editor/src/components/block-preview/index.js [hunk 3]
```diff
@@ -33,19 +37,57 @@ function BlockPreview( props ) {
...
+	const [ visibility, setVisibility ] = useState( 'hidden' );
+
+	let timerId;
+
+	// Dynamically calculate the scale factor
+	useLayoutEffect( () => {
+		const { clientId } = blocks;
+		timerId = setTimeout( () => {
+			const refNode = previewRef.current;
+			const containerWidth = refNode.offsetWidth - ( 14 * 2 );
```
Comment (getdave) [inline packages/block-editor/src/components/block-preview/index.js]: What are these magic numbers `14` and `2`?

File: packages/block-editor/src/utils/dom.js [hunk 4]
```diff
@@ -11,6 +11,15 @@ export function getBlockDOMNode( clientId ) {
 	return document.querySelector( '[data-block="' + clientId + '"]' );
 }
 
+export function getClockPreviewContainerDOMNode( clientId ) {
+	const domNode = getBlockDOMNode( clientId );
+	if ( ! domNode ) {
+		return;
+	}
+
+	return domNode.querySelector( '.wp-block-preview-container' );
```
Comment (youknowriad) [inline packages/block-editor/src/utils/dom.js]: I don't think we can rely on a a new `wp-block-preview-container` className added to all blocks.
Why not just `domNode.firstChild` (I'm pretty sure it would work consistently but we could fallback to `domNode` if it's not defined.

This ties the implementation of this function to the `BlockListBlock` component but we could prevent regressions by e2e test maybe.

File: packages/block-editor/src/components/block-preview/test/index.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * WordPress dependencies
+ */
+import { registerCoreBlocks } from '@wordpress/block-library';
+
+/**
+ * Internal dependencies
+ */
+import { BlockPreviewContent } from '../index';
+
+describe( 'BlockPreviewContent', () => {
```
Comment (getdave) [inline packages/block-editor/src/components/block-preview/test/index.js]: Thanks for starting some tests 👍

File: packages/block-editor/src/components/block-styles/style.scss [hunk 6]
```diff
@@ -12,6 +12,7 @@
 	overflow: hidden;
 	border-radius: $radius-round-rectangle;
 	padding: $grid-size-small * 1.5;
+	padding-top: calc(50% * 0.75 - #{ $grid-size-small } * 1.5);
```
Comment (getdave) [inline packages/block-editor/src/components/block-styles/style.scss]: Can you explain what this does?

File: packages/block-editor/src/components/block-preview/index.js (hunk 2 above)
Comment (youknowriad) [inline packages/block-editor/src/components/block-preview/index.js]: Why is this being cleared on a separate effect from the one that triggers it? should we move this to the previous effect?

File: packages/block-editor/src/utils/dom.js [hunk 7]
```diff
@@ -11,6 +11,15 @@ export function getBlockDOMNode( clientId ) {
 	return document.querySelector( '[data-block="' + clientId + '"]' );
 }
 
+export function getClockPreviewContainerDOMNode( clientId ) {
```
Comment (getdave) [inline packages/block-editor/src/utils/dom.js]: Is this a typo? Did you mean `getBlockPreviewContainerDOMNode` (note `Block` not `Clock`)? 😄

File: packages/block-editor/src/components/block-preview/index.js [hunk 8]
```diff
@@ -33,19 +37,57 @@ function BlockPreview( props ) {
...
+
+	const [ previewScale, setPreviewScale ] = useState( 1 );
+	const [ visibility, setVisibility ] = useState( 'hidden' );
+
+	let timerId;
+
+	// Dynamically calculate the scale factor
+	useLayoutEffect( () => {
+		const { clientId } = blocks;
+		timerId = setTimeout( () => {
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-preview/index.js]: Let's add a comment to explain why we need the setTimeout (the block list being rendered async..)

File: packages/block-editor/src/utils/dom.js (hunk 4 above)
Comment (getdave) [inline packages/block-editor/src/utils/dom.js]: You're relying on this selector being available on all Blocks. How can we guarantee that will be the case? Are we planning to add this selector to all Core Blocks? What about 3rd party Blocks.

This is the central issue at play here which I referenced in the original PR, namely that you cannot accurately and consistently target an element which wraps the _visual_ content of a Block.

I'm really interested to know if @youknowriad sees a way in which we could target consistently and what he feels about the approach outlined here.

**Review by getdave — DISMISSED** [review]
Great work on exploring this route again. You made some improvements over my original version of this. However I still think there is a fundamental flaw that is outside of your (our) control namely that we cannot reliably target the DOM element which represents the _**visual**_ width of the Block.

See https://github.com/WordPress/gutenberg/pull/16113#issuecomment-516897473

File: packages/block-editor/src/components/block-preview/index.js (hunk 2 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: I agree.

File: packages/block-editor/src/components/block-preview/index.js (hunk 8 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: It's pretty grim and I felt bad when I added this in my original PR. Is there no way to detect the `rendered` event?

Comment (youknowriad) [reply packages/block-editor/src/components/block-preview/index.js]: The solution I proposed (see the PR comment) solve some of it but I think the block list is rendered twice. First at init with an empty block list, and then a second time with the full block list. that's because of how `BlockEditorProvider` works. (init on mount). I think that's the main issue here. So when first rendered, the block list is empty.

File: packages/block-editor/src/utils/dom.js (hunk 4 above)
Comment (getdave) [reply packages/block-editor/src/utils/dom.js]: I will try your suggestion of `firstChild` to fully stress test this route.

File: packages/block-editor/src/components/block-preview/index.js [hunk 9]
```diff
@@ -33,19 +37,57 @@ function BlockPreview( props ) {
...
+			const refNode = previewRef.current;
+			const containerWidth = refNode.offsetWidth - ( 14 * 2 );
+
+			const previewDomElements = getClockPreviewContainerDOMNode( clientId );
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-preview/index.js]: I imagine that if we render multiple previews of the same block (same id), this could be ambiguious, should we add a ref to a parent to limit the scope of the selector?

File: packages/block-editor/src/components/block-preview/README.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+The unscaled dimension of the Block you are looking to preview. For example, if the Block looks good at `700x450` then you should set this value to `450`. See also `A note on source dimensions` above.
+
+### `scaleFactor`
+* **Type**: `number`
+* **Default**: `0.9`
+
+### `scaleToFit`
+* **Type**: `bool`
+* **Default**: `false`
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-preview/README.md]: This needs a refresh

File: packages/block-editor/src/components/block-preview/index.js [hunk 11]
```diff
@@ -1,20 +1,24 @@
 /**
  * External dependencies
  */
+
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-preview/index.js]:
```suggestion
```

File: packages/block-editor/src/components/block-preview/index.js (hunk 1 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-preview/index.js]: yes, I think if the blocks is an array (template), we should just scale to fit the whole width somehow.

File: packages/block-library/src/button/edit.js [hunk 12]
```diff
@@ -117,6 +117,7 @@ class ButtonEdit extends Component {
 					onChange={ ( value ) => setAttributes( { text: value } ) }
 					withoutInteractiveFormatting
 					className={ classnames(
+						'wp-block-preview-container',
```
Comment (youknowriad) [inline packages/block-library/src/button/edit.js]: we should revert these changes.

File: packages/block-editor/src/components/block-preview/README.md (hunk 10 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/README.md]: Agreed. I reckon we do a pass over all the docs once we know the API is solid.

File: packages/block-editor/src/components/block-preview/index.js (hunk 3 above)
Comment (retrofox) [reply packages/block-editor/src/components/block-preview/index.js]: Sorry, I removed the comments and forgot the clarify. We need to compute the offset of the container element. It's hardcoded for now.

File: packages/block-library/src/button/edit.js (hunk 12 above)
Comment (getdave) [reply packages/block-library/src/button/edit.js]: Just done this for all Blocks

File: packages/block-editor/src/components/block-preview/README.md (hunk 10 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/README.md]: I've added a `scope` param to the DOM util and passed in a suitable scope to limit the `querySelector`.

File: packages/block-editor/src/utils/dom.js (hunk 4 above)
Comment (getdave) [reply packages/block-editor/src/utils/dom.js]: I've resolved this now. We now just looked a the `firstChild`. It's not perfect but it's the best we can do.

File: packages/block-editor/src/utils/dom.js (hunk 7 above)
Comment (getdave) [reply packages/block-editor/src/utils/dom.js]: I've fixed this.

File: packages/block-editor/src/components/block-preview/index.js (hunk 9 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: I've fixed this by adding a `scope` param to the DOM util and passing a relevant scope into the call to get the Block contents.

File: packages/block-editor/src/components/block-preview/index.js (hunk 3 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: I've fixed this. Now dynamic.

File: packages/block-editor/src/components/block-preview/index.js (hunk 2 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: I've fixed this one.

File: packages/block-editor/src/components/block-preview/index.js (hunk 1 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: I've resolved this by looping all the Blocks. We'll need to watch out for perf concerns on this one though.

File: packages/block-editor/src/components/block-preview/index.js (hunk 8 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: Is there no way of propagating the change of state within `BlockEditorProvider` from "empty" to "I have some blocks" via a callback or something?

File: packages/block-editor/src/utils/dom.js (hunk 4 above)
Comment (getdave) [reply packages/block-editor/src/utils/dom.js]: I've implemented `firstChild`

**Comment by retrofox** [comment]
If we remove the idea behind of [querying the element to preview](https://github.com/WordPress/gutenberg/pull/16873/commits/b97475bcd5a68668034951ff2267b145dc58d7fc) I think it won't work at least in the way it was initially thought.

For instance, the `<Button />` childeNode represents the following:

Screen Shot 2019-08-02 at 8 03 43 AM

While the element that we use to compute the scale is the next one:

image

**Comment by getdave** [comment]
> If we remove the idea behind of querying the element to preview I think it won't work at least in the way it was initially thought.

There is no alternative but to remove it. We hit that roadblock already on https://github.com/WordPress/gutenberg/pull/1611. We cannot control the DOM of Blocks. We just don't have that control.

The best we can do is Riad's suggestion of `firstChild` which is ok, but will never be perfect.

**Comment by youknowriad** [comment]
> For instance, the <Button /> childeNode represents the following:

I think for the button we can solve this by making that container "div" take the size of its content. (especially for the "unselected" state)

**Comment by retrofox** [comment]
> > For instance, the childeNode represents the following:

> I think for the button we can solve this by making that container "div" take the size of its content. (especially for the "unselected" state)

But that's the problem, Raid. We make that assumption for the `<Block />` element, but for the `<Table />` or `<Pullquote />` we should pick up the `<figure>` element. There isn't consistency in terms of the tree of a Block, which is great but not useful for this purpose.

**Comment by youknowriad** [comment]
I think there's two things. The fact that the wrapper of the button doesn't take its container's size can be considered a bug IMO because that's not how the frontend looks and the editor should match the frontend for blocks.

The other aspect here is that depending on the content of the block, the preview might not look great. (Say you type just a few words in a quote compared to a full line text). For this case, a solution (that is needed elsewhere as well) is to add a new API to blocks to be able to define "an example content" instead of the content shown in the block while opening the block switcher. (that should be a separate PR though)

**Comment by youknowriad** [comment]
Actually, I just noticed that there's a full-width div for the button block in frontend too. you can argue that in that case, it's "normal" that the button block shows up as very small because its content is the whole div (what if it has a background image defined by a theme for instance, as silly as it is, it's part of the content). 

This is to say IMO, we should tweak the button block to make its container take the size of its content in both editor and frontend.

**Comment by retrofox** [comment]
I think we can try to make it work. Among all options in the developing flow, we considered just preview a handful of blocks. So this scope fits this option.

And also, we could improve the `getBlockPreviewContainerDOMNode()` in order to try to return the best (less bad) element to preview. In the worse of cases, we can return the childNode:

```es6
export function getBlockPreviewContainerDOMNode( clientId ) {
	const domNode = getBlockDOMNode( clientId );

	if ( ! domNode ) {
		return;
	}

	const previewNode = domNode.querySelector( '.wp-block-preview-container' );
	if ( previewNode ) {
		return previewNode;
	}

	return domNode.firstChild || domNode;
}
```

Maybe we could whitelist some classes as well?

Finally, is it too crazy suggest to the dev community using this method or something similar to get a nicer preview?

**Comment by youknowriad** [comment]
> Finally, is it too crazy suggest to the dev community using this method or something similar to get a nicer preview?

It's not but I think conceptually speaking. The content of that block is the whole div (first-child) and that seems to me like a consistent answer we can give the community too.

> Q: Why the preview of my block is showing very small?
> R: it's not, we're just scaling the preview to show the whole content of your block including the wrapper?

We could consider adding a new "className" as API in the future depending on the feedback?

**Comment by retrofox** [comment]
> > Q: Why the preview of my block is showing very small?
>
> > R: it's not, we're just scaling the preview to show the whole content of your block including the wrapper?

That sounds very good. 

> We could consider adding a new "className" as API in the future depending on the feedback?

Yes, it's a matter to be consistent and get consensus in defining the structure of a block.

> the content of that block is the whole div (first-child) and that seems to me like a consistent answer we can give the community too.

I'd like to go for this rule as the main one. I don't like using the CSS class but it seemed to be the less bad. Now, if we agree to tweak the core blocks is an option then better than better. AFAIK, they are only five blocks that are using the preview.

**Comment by getdave** [comment]
Ok the new objectives are:

* Fix Core Blocks so that their wrapping elements reflect their visual width (@retrofox to make a list)
* Write updates guidelines for Block authors advising to ensure their Blocks reflect their visual width

Then we can look to ship this PR.

I also think we need to reintroduce an option to render the preview at a set aspect ratio. Sometimes scaling to the widest Block isn't what you want, especially if there is more than 1 Block.

**Comment by retrofox** [comment]
Core blocks list:

- [x] button
- [x] separation
- [x] table
- [x] quote
- [x] pullquote

**Comment by getdave** [comment]
> Core blocks list:
> 
> * [x]  button
> * [x]  separation
> * [x]  table
> * [x]  quote
> * [x]  pullquote

Great work 👏

**Comment by getdave** [comment]
I've created https://github.com/WordPress/gutenberg/pull/16885 which allow us to hide the block appender within previews.

Needs review.

File: packages/block-editor/src/components/block-styles/index.js [hunk 13]
```diff
@@ -116,17 +116,13 @@ function BlockStyles( {
 							}
 						} }
 						onMouseEnter={ () => onHoverClassName( styleClassName ) }
-						onMouseLeave={ () => onHoverClassName( null ) }
+						// onMouseLeave={ () => onHoverClassName( null ) }
```
Comment (getdave) [inline packages/block-editor/src/components/block-styles/index.js]: @retrofox Was this intentional?

Comment (retrofox) [reply packages/block-editor/src/components/block-styles/index.js]: testing purpose. remoiving.

File: packages/block-editor/src/components/block-preview/style.scss [hunk 14]
```diff
@@ -9,38 +10,64 @@
...
-			display: none;
-		}
+	.block-editor-block-preview__title {
+		margin-bottom: 10px;
+		color: $dark-gray-300;
 	}
 }
 
+// These rules ensure the preview scales smoothly regardless of the container size.
+.editor-block-preview__container {
```
Comment (getdave) [inline packages/block-editor/src/components/block-preview/style.scss]: I don't think we should be using the `.editor-` namespace. It's inside the `block-editor` package.

**Comment by obenland** [comment]
#### Quotes
Quotes with only a few words are being displayed rather small, in both thumbnail and preview. Initially I assumed it was because of the width of the content area, but even within a column block where the content width would be 1/3, it's still shown the same way.

image

#### Buttons
The issue where a button wraps lines in the content but not the thumbnail or preview persists. It's probably enough of an edge case that we could just live with it, but I thought I'd still bring it up.

image

#### Tables

Tables with lots of cells are bleeding out the top of the thumbnails and previews. I think it's fine to not show the entire table, but let's make sure we always see its top, and it overflows at the bottom.

image

#### Pullquotes

Pullquotes pretty much only look good in thumbnail and preview, when they're not pulled :) 
When they're aligned, they get cut off in thumbnail and preview, and have odd line breaks when they're full width or wide width

**Comment by jasmussen** [comment]
Impressive work here, thank you folks for running with the original PR and taking it in a new direction that from a glance looks much better already.

It seems like the primary remaining issue is that the preview now works badly with specific instances of content, is that correct? Or are there additional challenges?

Apologies in advance if I wade in to this discussion like a bull in a china shop, I've just had some much needed vacation. So strawman ideas:

- Could we on a per-block basis, provide some "preview best practices", like setting a max-width on the button ONLY in the preview, and elide the text if it's beyond that? That way it would be up to each block to ensure a good preview.
- Can we provide pre-made example text on a per-block basis so we know the previews always look good? Or is the necessary expansion of the block API surface undesirable?

**Comment by retrofox** [comment]
> Initially I assumed it was because of the width of the content area, but even within a column block where the content width would be 1/3, it's still shown the same way.

It's because of the width of the `quote` always takes 100%, so beyond what's the width of the user sees in the screen, it will try fix always to 100% of its container. It happens for all blocks, and for some of them, the result is not the ideal. For this reason, we've slightly talked about an example component. 

IMO, what we can do now is adjust the preview to makes it look as best as possible without increasing the complexity of this approach.

**Comment by retrofox** [comment]
> Could we on a per-block basis, provide some "preview best practices", like setting a max-width on the button ONLY in the preview, and elide the text if it's beyond that? That way it would be up to each block to ensure a good preview.

Yes !!!, something that [we considered last week](https://github.com/WordPress/gutenberg/pull/16873#issuecomment-517682030): 

> Write updates guidelines for Block authors advising to ensure their Blocks reflect their visual width


> Can we provide pre-made example text on a per-block basis so we know the previews always look good? Or is the necessary expansion of the block API surface undesirable?

Yes too! :-D
That's something that @youknowriad suggested.

File: packages/block-editor/src/components/block-preview/style.scss (hunk 14 above)
Comment (marekhrabe) [reply packages/block-editor/src/components/block-preview/style.scss]: correct, everything new would start with `.block-editor-block-preivew…`. some old `.editor-…` were kept for backwards compatibility but we will not be adding any new

**Comment by obenland** [comment]
@youknowriad I'd love for you to take another look at this :)

The reason for the long buttons to not preview line breaks correctly is that allowing buttons to wrap whitespace is breaking short buttons with spaces in them. It's probably worthwhile to investigate that further, I'd consider it rather an edge case however and would suggest to not let it hold up this PR.

I took another look at vertical alignments of blocks that are smaller than the preview window and pushed some fixes for that. It brought back a non-optimal experience in the large preview where it aligns all blocks at the top. The preview window hight is set to `0` by the time we calculate the relative height of the block, and I'm not sure why that is or how to reconcile that.

Both of these caveats seem rather minor and can be iterated on. How do you feel about merging what we have so far and give those two things another look after?

File: packages/block-library/src/button/editor.scss [hunk 15]
```diff
@@ -91,3 +75,21 @@
 		margin-top: $grid-size-large;
 	}
 }
+
+// Tweak styles for the Preview.
+.block-editor-block-preview__content {
```
Comment (youknowriad) [inline packages/block-library/src/button/editor.scss]: I don't think we should be having styles per block for previews. I think the preview should be generic enough. The problem with these styles is that the block can be nested in templates and these templates can be previewed too. I'm pretty sure this would break the previews of the templates using buttons.

File: packages/block-editor/src/components/block-preview/index.js [hunk 16]
```diff
@@ -9,38 +9,108 @@ import classnames from 'classnames';
...
+			const previewContainer = previewRef.current;
+
+			if ( ! previewContainer ) {
+				return;
+			}
+
+			const comparisonBlock = blockClientIds.reduce( ( acc, currClientId ) => {
+				// Selector scoped to `previewContainerDomElement` to avoid global selector being ambiguous in the case
+				// of multiple previews on the same view
+				const previewDomElement = getBlockPreviewContainerDOMNode( currClientId, previewContainer );
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-preview/index.js]: I guess this is done to pick the biggest block. I feel it's not needed. I think it could be as simple as saying, if we have more than one block (previewing templates), just pick the full width (700 right now)

File: packages/block-library/src/button/editor.scss (hunk 15 above)
Comment (retrofox) [reply packages/block-library/src/button/editor.scss]: What I'm trying to do here is applying styles only when the blocks are being previewed. The styles defined below are needed to adjust the preview for the button block. Without them, the preview gets weird positions, especially when it sets the button alignment.

If I move these rules out of the `.block-editor-block-preview__content` scope, the styles will be applied to all places where the blocks are inserted, which is undesired.

> The problem with these styles is that the block can be nested in templates and these templates can be previewed too.

If I'm not wrong, these templates always will contain the `.block-editor-block-preview__content` css class into its content, so it means that these styles will be always applied, as long as the blocks are previewed using the `<BlockPreview />` block.

File: packages/block-editor/src/components/block-preview/index.js (hunk 16 above)
Comment (retrofox) [reply packages/block-editor/src/components/block-preview/index.js]: 👍 Ok. in that case we need to find a dynamic way to define the width? We thought using a component property, let's say for instance `viewportWidth`, or something like that.

File: packages/block-library/src/button/editor.scss (hunk 15 above)
Comment (youknowriad) [reply packages/block-library/src/button/editor.scss]: But what I'm saying is that they probably shouldn't be applied when previewing templates. They should only be applied when previewing a single button block.

Also, I feel like this is not a generic enough fix. Imagine a small third party block, will we force them to have special styles for the preview?

Isn't there any way to fix the "position" of the scaled preview based on the scaled element? Right now we're trying to center horizontally, even aligning left would be better (it won't solve right aligned blocks though that's why I think the ideal solution is just to align the "left" of the preview with the "left" of the scaled element)

Comment (retrofox) [reply packages/block-library/src/button/editor.scss]: > But what I'm saying is that they probably shouldn't be applied when previewing templates. They should only be applied when previewing a single button block.

👍

Comment (retrofox) [reply packages/block-library/src/button/editor.scss]: > Isn't there any way to fix the "position" of the scaled preview based on the scaled element?

it seems to be kind tricky if we want to keep all the variants: button text (size), alignments, etc.
What do you think reducing the scope of these rules only into the styles preview?

Comment (youknowriad) [reply packages/block-library/src/button/editor.scss]: Not sur eI follow why it's tricky. Here's some pseudo code to understand my idea.

```js
const scaledElementRect = element.getBoundingClientRect(); // element is the scaled  element, the first child of the block
const rootOfPreviewRect = previewRoot.getBoundingClientRect(); // previewRoot is the root element of the BlockList

const diff = scaledElementRect.left - rootOfPreviewRect.left; 
// And then apply this diff as a CSS transform or left property of the preview
````

File: packages/block-editor/src/components/block-preview/index.js [hunk 17]
```diff
@@ -49,27 +48,20 @@ export function BlockPreview( {
...
-
-				return acc;
-			}, { height: 0, width: 0 } );
-
-			const scale = previewContainer.offsetWidth / comparisonBlock.width || 1;
-
-			setIsTallPreview( comparisonBlock.height > previewContainer.offsetHeight );
-			setPreviewScale( scale );
-			setVisibility( 'visible' );
+				setIsTallPreview( previewElementHeight > previewElementHeight );
```
Comment (obenland) [inline packages/block-editor/src/components/block-preview/index.js]: This will always be `false`. The second `previewElementHeight` should be a `containerElementHeight`

**Comment by obenland** [comment]
The "Build artifacts" test keeps failing, not sure if that is something that's caused by this PR?

**Comment by youknowriad** [comment]
@obenland That's one of two things `npm run docs:build` (most likely to update the docs). or `npm install` to update the lock file.

**Review by obenland — APPROVED** [review]

**Comment by youknowriad** [comment]
What a team effort 🎉

**Comment by obenland** [comment]
Thank you SO much for all your help!

**Comment by shaunandrews** [comment]
💋

**Comment by jasmussen** [comment]
HOOLY wow. 👏👏👏

**Comment by getdave** [comment]
👏 👏 👏 👏

File: packages/block-library/src/button/editor.scss [hunk 18]
```diff
@@ -91,3 +75,7 @@
 		margin-top: $grid-size-large;
 	}
 }
+
+.wp-block-button-wrapper {
+	display: table;
```
Comment (talldan) [inline packages/block-library/src/button/editor.scss]: Is `display: table;` the only way to solve this? It seems like a bit of a hack that'll make life a lot harder in the long run.

File: packages/block-editor/src/components/block-preview/style.scss [hunk 19]
```diff
@@ -1,34 +1,87 @@
...
+	position: absolute;
+
+	// Vertical alignment. It works with the transform: translate(-50%, -50%)`,
+	top: 0;
+	left: 0;
+
+	// Important to set the origin.
+	transform-origin: top left;
+
+	// Resetting paddings, margins, and other.
```
Comment (aduth) [inline packages/block-editor/src/components/block-preview/style.scss]: I'm not sure if it was introduced here, but while the preview shows well in most contexts, in the block transform menu, I see horizontal scroll:

image

From some inspection, it seems to be because of the negative offsets from pseudo-elements surrounding the block in the block list:

https://github.com/WordPress/gutenberg/blob/be671eaaec240b51b92110f68c0f03bcafe3842a/packages/block-editor/src/components/block-list/style.scss#L1120-L1122

I'd think that ideally, we could have some concept of a "read-only" block list rendering, since these previews have no use for all of the extra control elements of a block.

But I comment here because it seems like you had to deal with this anyways. Maybe something we can be addressing with `overflow: hidden` ? 

(Aside: I find it helps to use the "Show Scroll Bars: Always" option in macOS, to help discover these sorts of issues: https://cloudup.com/cqF641QROEe)

Comment (youknowriad) [reply packages/block-editor/src/components/block-preview/style.scss]: I don't have a good answer here, This was a very complex PR and I know that it can be tricky to get right on all situations.

> read-only mode or preview mode

I see this mentioned over and over and I think we'll probably need that for several things (previews and also some special FSE mode where some blocks are "enabled" while others are not)

> negative left/right

This looks specific to the hover and selected styles, so yeah if we can move this out of the flow somehow (overflow: hidden maybe but I'd be very cautious), it could work or just remove the selected/hover styles entirely in preview.

Comment (retrofox) [reply packages/block-editor/src/components/block-preview/style.scss]: 👋 Honestly, I don't remember exactly the purposes of introducing these resets here. As Riad mentioned, it's a very complex PR and any modification can affect other instances where the preview is being used. AMost of these resets are usually an attempt to make cross-theme support. 

`read-only`, `view`, etc. have been talked as well. I thought that even we could remove all elements which are not needed just for the preview (appender, drop-zone, etc) in order to make the preview process more efficient.
